util/device_sw_utils/extract_sw_logs.py

```diff
@@ -0,0 +1,123 @@
+given section of the ELF file generated for the device sw application. It appends
+the offset address of each string constant with the base address of the section
+to get the compete address of the log. The SW log monitor looks up this database
+to extract the logs at the right address and prints it with the appropriate severity.
+"""
+
+import argparse
+import logging as log
+import os
+import re
+import subprocess
+import sys
+
+
+def extract_sw_logs(readelf, section, input, output):
+    # Get the base address of the section
+    addr = 0
+    cmd = "readelf -S " + input + " | grep -F " + section
+    cmd_output = subprocess.run(cmd,
+                                check=True,
+                                shell=True,
+                                stdout=subprocess.PIPE,
+                                stderr=subprocess.STDOUT)
+    section_attrs = cmd_output.stdout.decode('utf-8')
+    # Output is of thie format:
+    # [10] .logs             PROGBITS        00000000 005000 000149 00 ...
+    # The start addr is the first attribute after 'PROGBITS'.
+    match = re.match(r"\s*\[.*\]\s+" + section + \
+                     "\s+\w+\s+([a-fA-f0-9]+)\s+.*", section_attrs)
+    if match:
+        addr = int(match.groups()[0].strip(), 16)
+    else:
+        log.info(
+            "Failed to extract the start addr of \"%s\" in \"%s\" file."
+            "Using addr = 0", section, input)
+
+    # Extract the logs from the section using the readelf utility.
+    cmd = [readelf, '-p', section, input]
+    cmd_output = subprocess.run(cmd,
+                                check=True,
+                                stdout=subprocess.PIPE,
+                                stderr=subprocess.STDOUT)
+    # Decode log texts into utf-8 string, replace bytes that do not conform.
+    text = cmd_output.stdout.decode('utf-8',
+                                    errors='replace').strip().split("\n")
+
+    # Each line of the log looks like this:
+    #   [   a4] Hello World!
+    #   OR
+    #   [   a4] I "hello_world.c" 25 0 "Hello World!"
+    #
+    # Convert this to:
+    # addr: 900000a4
+    # severity: I
+    # file: hello_world.c (or unknown_file)
+    # line: 25 (or 0)
+    # nargs: 0
+    # format: Hello World
+    result = ""
+    for line in text:
+        match = re.findall(r"^\s*\[\s*([a-fA-F0-9]+)\]  (.*)$", line)
+        if match:
+            subst_list = {}
+            for item in match:
+                offset_val = addr + int(item[0], 16)
+                log = item[1]
+                # Get rid of the ^M and ^J characters (cr, lf).
+                log = log.replace("^M", "")
+                log = log.replace("^J", "")
+                # Convert into 'address: log' format.
+                result += "addr: {}\n".format(hex(offset_val)[2:])
+                # Parse log into additional information (if available).
+                format_match = re.match(
```

Comment:
All of this code would go well into a separate function to ease testing and make the code more understandable.